### ISSUE TITLE
The trouble with telnets

### ISSUE THREAD

**Issue by Rhialto** [simh/simh#177]
I have been trying to get a Kermit-11 working on an RSX-11M+ 4.6
installation.  One of the steps in that process is to use kermit to have
it type a hexdump into a file. So I use Kermit's telnet command to
connect to one of RSX's serial lines, type CREATE K11RSX.HEX, escape
back to my local host, and use Kermit's TRANSFER command to make it type
the hexdump.

This report is on the TELNET step of the procedure, which won't work.
Kermit's telnet client sends some IAC WILL options, which never get a
reply.
(Kermit suggests to use a /nowait switch to paper over that, which also
doesn't work, but that seems a separate issue.)
Kermit rightly expects some reply to its IAC WILLs.
The solution would be to simply send IAC WONT &lt;option> or IAC DONT &lt;option; for any option
simh doesn't recognize (while preventing infinite DO/WONT or WILL/DONT (etc) loops).

To compare:

Connecting with (NetBSD 6.1.5) telnet to the PDP-11, with options debugging:
```
$ telnet
telnet> set debug
Will turn on socket level debugging.
telnet> set options
Will show option processing.
telnet> set netdata
Will print hexadecimal representation of network traffic.
telnet> set prettydump
Will print user readable output for "netdata".

telnet> open localhost 2312
Trying ::1...
Connected to localhost.
Escape character is '^]'.
< 0x0   *ff fb 22*ff fb 03*ff fb 01*ff fb 00*ff fd 00
RCVD WILL LINEMODE
SENT DONT LINEMODE
RCVD WILL SUPPRESS GO AHEAD
SENT DO SUPPRESS GO AHEAD
RCVD WILL ECHO
SENT DO ECHO
RCVD WILL BINARY
SENT DO BINARY
RCVD DO BINARY
SENT WILL BINARY
> 0x0   *ff fe 22*ff fd 03*ff fd 01*ff fd 00*ff fb 00
< 0x0    0a 0d 0a 43 6f 6e 6e 65 63 74 65 64 20 74 6f 20
< 0x10   74 68 65 20 50 44 50 2d 31 31 20 73 69 6d 75 6c
< 0x20   61 74 6f 72 20 56 48 20 64 65 76 69 63 65 2c 20
< 0x30   6c 69 6e 65 20 31 0d 0a 0a


Connected to the PDP-11 simulator VH device, line 1

> 0x0    0d
< 0x0    0a

< 0x0    0d 0d 0a 3e

>
```
and With Kermit:
```
$ kermit
C-Kermit 9.0.302 OPEN SOURCE:, 20 Aug 2011, for NetBSD 6.0.1+SSL (64-bit)
 Copyright (C) 1985, 2011,
  Trustees of Columbia University in the City of New York.
Type ? or HELP for help.
C-Kermit> set debug session
C-Kermit> telnet localhost 2312
 DNS Lookup...  Trying 127.0.0.1...  Reverse DNS Lookup... (OK)
TELNET SENT WILL AUTHENTICATION
TELNET SENT WILL NAWS
TELNET SENT WILL TERMINAL-TYPE
TELNET SENT WILL NEW-ENVIRONMENT
TELNET SENT WILL COM-PORT-CONTROL
<wait for outstanding negotiations>
TELNET RCVD WILL LINEMODE
TELNET SENT DONT LINEMODE
TELNET RCVD WILL SUPPRESS-GO-AHEAD
TELNET SENT DO SUPPRESS-GO-AHEAD
TELNET RCVD WILL ECHO
TELNET SENT DO ECHO
TELNET RCVD WILL BINARY
TELNET SENT DO BINARY
TELNET RCVD DO BINARY
TELNET SENT WILL BINARY
 Negotiations...............................
*************************
The Telnet server is not sending required responses.

?Telnet waiting for response to WILL TERMINAL-TYPE
?Telnet waiting for response to WILL NAWS
?Telnet waiting for response to WILL AUTHENTICATION
?Telnet waiting for response to WILL NEW-ENVIRONMENT
?Telnet waiting for response to WILL COM-PORT-CONTROL

You can continue to wait or you can cancel with Ctrl-C.
In case the Telnet server never responds as required,
you can try connecting to this host with TELNET /NOWAIT.
Use SET HINTS OFF to suppress further hints.
*************************
```
Waiting longer at this point still never completes, since there are never any IAC WILL or IAC WONT replies for these options.

[edited for escaping less-than]

**Comment by markpizz** [comment]
One interpretation of things might suggest that since Kermit has just agreed to enter binary mode, nothing else should be interesting...  The existing 'pseudo negotiation' has been sufficient for quite a while.  It is achieved by merely blurting out a static string and just digesting (if binary mode is acknowledged) or ignoring what comes back (i.e. no round trips working out details).  

If the simh code were to change, what response is Kermit expecting for these options that the simh side isn't interested in?  Is "DONT option" right, or is "WONT option" right?

**Comment by Rhialto** [comment]
On Tue 20 Jan 2015 at 20:52:38 -0800, Mark Pizzolato wrote:
> One interpretation of things might suggest that since Kermit has just agreed to enter binary mode, nothing else should be interesting...  The existing 'pseudo negotiation' has been sufficient for quite a while.  It is achieved by merely blurting out a static string and just digesting (if binary mode is acknowledged) or ignoring what comes back (i.e. no round trips working out details).  

I've been re-reading the Telnet RFC at
http://www.rfc-editor.org/rfc/rfc854.txt, in particular pages 2-4. It
seems that the formal MAY, SHOULD and MUST terminology was not
introduced yet at the time, so how compulsory the written description
is, is slightly inclear. In any case, I feel that both sides are at
fault here (its /NOWAIT option seems not to work), and I will report
that bug to the Kermit project.

That being said, the text suggest there are cases where an option should
not be responded to, to prevent loops, However I don't think that
applies here, because simh doesn't know any of the outstanding options
(TERMINAL-TYPE, NAWS, AUTHENTICATION, NEW-ENVIRONMENT and
COM-PORT-CONTROL) and can't handle it if Kermit were to go ahead and
assume they are enabled.

> If the simh code were to change, what response is Kermit expecting for these options that the simh side isn't interested in?  Is "DONT option" right, or is "WONT option" right?

I read the text as saying that if one side initiates with a WILL/WONT
offer, then the other side should say DO/DONT if it doesn't like the
option.

Although it may be normal to initiate with a DO/DONT command, and the
other side responds with WILL/WONT.

However there is te loop-prevention from paragraph 3 on page 2, which
basically says that offers of/requests for an option should only be sent
if this represents a change. That seems to be the case here, so Kermit's
WILLs should be responded to with DONT, and I think that Kermit should
acknowledge that change with WONT (which should not be replied to).

-Olaf.
-- 
___ Olaf 'Rhialto' Seibert  -- The Doctor: No, 'eureka' is Greek for
\X/ rhialto/at/xs4all.nl    -- 'this bath is too hot.'

**Comment by markpizz** [comment]
Something which I neglected to say at the beginning of this was that the existing recipe has been working with only two complaints since the beginning.  Any number of telnet clients have successfully communicated.  There is no specific list, but changing things here might break something else which I have no specific way to test since I don't know the population of telnet clients.

The two complaints were:
1) connections from the PuTTY client on Windows sent a sequence of options, not all of which were specifically recognized as telnet options.  The original code passed these un-recognized options through as data in the session which disturbed some simulators.  Clearly this was wrong on the simh side so the implementation was changed to separate and ignore any inband option negotiation traffic and not pass unknown options as data.  No change on the wire, but a fix for the simulated application. Now PuTTY connects cleanly.
2) The Windows command line telnet client is fundamentally broken.  It is broken in a subtle way but that issue has existed since the earliest Microsoft client.  What is going on is that most telnet servers don't actually configure the session into binary mode (while simh does).  The Windows telnet client cooperates AND AGREES to enter binary mode.  The key distinction in binary mode is that CR is now supposed to be sent as a bare CR character and not the CRLF sequence which is required in ASCII mode.  The Microsoft client ALWAYS sends CRLF without regard to the state of binary mode.  After weeks of conversations with Microsoft's support folks, they came back with some screwy 'work arounds':  1) You can make a registry setting to cause CR to always be sent bare without regard to the binary mode status (this just breaks something else).  OR 2) you can drop the telnet client to command mode and enter a command which will cause CR to be sent bare for the rest of the current telnet session.  Clearly both solutions are ridiculous and do not address the fact that NO user will ever do this and the fact that the application is broken.  Since they were able to provide these wonderful suggestions they declared that the there isn't a problem so no fix is appropriate.  The right fix for users is to use a different telnet client.

Back to the Kermit problem...

If I change the implementation to send a "DONT option" in response to all "WILL option" requests which come in I end up with the following:

     (/home/mark/) C-Kermit>set debug session
     (/home/mark/) C-Kermit>telnet localhost:1234
      DNS Lookup...  Trying 127.0.0.1...  Reverse DNS Lookup... (OK)
     TELNET SENT WILL AUTHENTICATION
     TELNET SENT WILL NAWS
     TELNET SENT WILL TERMINAL-TYPE
     TELNET SENT WILL NEW-ENVIRONMENT
     TELNET SENT WILL COM-PORT-CONTROL
     <wait for outstanding negotiations>
     TELNET RCVD WILL LINEMODE
     TELNET SENT DONT LINEMODE
     TELNET RCVD WILL SUPPRESS-GO-AHEAD
     TELNET SENT DO SUPPRESS-GO-AHEAD
     TELNET RCVD WILL ECHO
     TELNET SENT DO ECHO
     TELNET RCVD WILL BINARY
     TELNET SENT DO BINARY
     TELNET RCVD DO BINARY
     TELNET SENT WILL BINARY
     TELNET RCVD DONT AUTHENTICATION
     TELNET RCVD DONT NAWS
     TELNET RCVD DONT TERMINAL-TYPE
     TELNET RCVD DONT NEW-ENVIRONMENT
     TELNET RCVD DONT COM-PORT-CONTROL
     <no outstanding negotiations>
     TELNET SENT DO KERMIT
     TELNET SENT WILL KERMIT
     <wait for outstanding negotiations>
     TELNET RCVD DONT KERMIT
      Negotiations............................
     *************************
     The Telnet server is not sending required responses.

So, now I guess that responding to "DO option" for uninteresting options should return "WONT option"...

Of course, keeping track of the fact that the "DONT option" or "WONT option" had been sent and only responding as described one time....

Sound reasonable?

**Comment by markpizz** [comment]
I've committed the above changes.  Please re-open this if you see any furthur issues.

**Comment by Rhialto** [comment]
Great, thanks! I've tested it and it works.

But just a small thing: you test for a long list of uninteresting
options. Would it not be shorter (and more resilient) to check for the
short list of interesting options (and negate)?

I'm getting a bit further with my RSX-side Kermit adventures: I have a
K11RSX.TSK executable[*], but it seems to hang when receiving files. And
the ABOrt command seems to be missing. Nobody has an idea where to find
it?

Also, Kermit has a bug where this TELNET /NOWAIT option seems to be
ignored unless you already specify it the first time you use TELNET.
I'll probably report that there. There is a bugfix related to options in
Kermit 9.0.304 Dev.05 (according to http://www.kermitproject.org/ckdaily.html)
but I'll need to check if that would have helped here.

Hopefully this reply doesn't re-open the issue, otherwise I'll just
close it again.

[*] obtained previously with some workarounds, which had their own
problems, which may or may not be fixed by this telnet fix. If unfixed
I'll report those separately.

**Comment by markpizz** [comment]
I thought of the inverse test, but I was in a rush and didn't have time to work out and test the negative list.  This works and could always be changed later.
